library/common/jni/jni_interface.cc

```diff
@@ -133,6 +135,10 @@ static void* jvm_on_headers(const char* method, envoy_headers headers, bool end_
                                          end_stream ? JNI_TRUE : JNI_FALSE);
 
   env->DeleteLocalRef(jcls_JvmCallbackContext);
+  if (end_stream) {
+    env->DeleteGlobalRef(j_context);
+  }
+
```

Comment:
Exactly one of `on_error`, `on_complete`, or `on_cancel` should be called for every stream. Those should be the only place we delete `j_context` like this.

Reply:
Opened https://github.com/lyft/envoy-mobile/pull/1089

Reply:
Oh right, I've forgotten we have an `on_complete` which is called after a stream successfully finishes